File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#
class Solution:
    def longestPalindrome(self, s: str) -> str:
        size = len(s)
        if size <= 1:
            return s
        # 二维 dp 问题
        # 状态：dp[l,r]: s[l:r] 包括 l，r ，表示的字符串是不是回文串
        # 设置为 None 是为了方便调试，看清楚代码执行流程
        dp = [[False for _ in range(size)] for _ in range(size)]

        longest_l = 1
        res = s[0]

        # 因为只有 1 个字符的情况在最开始做了判断
        # 左边界一定要比右边界小，因此右边界从 1 开始
        for r in range(1, size):
            for l in range(r):
                # 状态转移方程：如果头尾字符相等并且中间也是回文
                # 在头尾字符相等的前提下，如果收缩以后不构成区间（最多只有 1 个元素），直接返回 True 即可
                # 否则要继续看收缩以后的区间的回文性
                # 重点理解 or 的短路性质在这里的作用
                if s[l] == s[r] and (r - l <= 2 or dp[l + 1][r - 1]):
                    dp[l][r] = True
                    cur_len = r - l + 1
                    if cur_len > longest_l:
                        longest_l = cur_len
                        res = s[l:r + 1]
            # 调试语句
            # for item in dp:
            #     print(item)
            # print('---')
        return res
''''这是另一种思路的动态规划方案，与c++中的动态规划实现方案不同的是，这种思路采用了
统一化的思想，同样是滚雪球，将所有的状态都置为false，首先从最左端开始，只要可以使两端相等且中间只有一个数
或者没有数，即可判断该串是回文串，而后右界逐渐向右扩张，这种算法稳定的根源在于，想要判断一个
串是否为回文串 就要判断他的左右各进一位的子串是否为回文串，而这个工作在之前的滚雪球中已经完成了，因为
滚雪球是从小滚到大的，所以小的子串的状态是先于大的子串确定的。

File: 5.longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.length()==1) return s;//大小为1的字符串必为回文串
        string rev=s;//rev存放s反转结果
        string res;//存放结果
        std::reverse(rev.begin(),rev.end());
        if(rev==s) return s;
        int len=0;//存放回文子串的长度
        for(int i=0;i<s.length();i++)//查找s与rev的最长公共子串
        {
            string temp;//存放待验证子串
            for(int j=i;j<s.length();j++)
            {
                temp=temp+s[j];
                if(len>=temp.length())
                    continue;
                else if(rev.find(temp)!=-1)//在rev中找到temp
                {
                    string q=temp;//q用来验证temp是否是回文子串
                    std::reverse(q.begin(),q.end());
                    if(q==temp)
                    {
                        len=temp.length();
                        res=temp;
                    }
                }
                else break;
            }
            temp="";
        }
        return res;
    }
};
/*方法一 倒置字符串法 将原字符串倒置 然后逐位寻找公共子串 ，如果存在公共子串 则需要
再加一个判断 判断其本身是不是回文串  所以其实在判断过程中是比较繁琐的 也占用的较多的
时间 所以最好的方法依然是动态规划法*/

class Solution {
public:
    string longestPalindrome(string s) {
        int len=s.size();
        if(len==0||len==1)
            return s;
        int start=0;//回文串起始位置
        int max=1;//回文串最大长度
        vector<vector<int>>  dp(len,vector<int>(len));//定义二维动态数组
        for(int i=0;i<len;i++)//初始化状态
        {
            dp[i][i]=1;
            if(i<len-1&&s[i]==s[i+1])
            {
                dp[i][i+1]=1;
                max=2;
                start=i;
            }
        }
        for(int l=3;l<=len;l++)//l表示检索的子串长度，等于3表示先检索长度为3的子串
        {
            for(int i=0;i+l-1<len;i++)
            {
                int j=l+i-1;//终止字符位置
                if(s[i]==s[j]&&dp[i+1][j-1]==1)//状态转移
                {
                    dp[i][j]=1;
                    start=i;
                    max=l;
                }
            }
        }
        return s.substr(start,max);//获取最长回文子串
    }
};
/*详细的分析一下这道题的思路 程序分为以下两步进行，首先我们可以确定的是 “最短的”回文
字符串：单个的字符 都是回文字符串 其次连续两个相同的字符都是回文字符串；
然后，开始从长度为3的字符串开始检索，首先设定长度为3 然后从数列的第一个开始遍历，遍历一遍
以后 ，所有长度为3的子串就全部确定是否为回文串了（注意第一步已经将单个字符的状态全部确定），
然后长度增加为4，遍历（注意第一步已经 将两个字符的状态全部确定），而后就可以通过已经确定的
长度为1,2,3,4的状态，进一步确定后面的所有长度的状态字，最后找到长度最长的。
